### PR TITLE
CLOUDP-114248: Store all config files with mongocli folder

### PR DESCRIPTION
**Pull request by blva**
## Proposed changes

_Jira ticket:_ [CLOUDP-114366](https://jira.mongodb.org/browse/CLOUDP-114366)

## Checklist

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation in document requirements section listed in [CONTRIBUTING.md](https://github.com/mongodb/mongocli/blob/master/CONTRIBUTING.md) (if appropriate)
- [x] I have addressed the @mongodb/docs-cloud-team comments (if appropriate)
- [x] I have updated [e2e/E2E-TESTS.md](https://github.com/mongodb/mongocli/blob/master/e2e/E2E-TESTS.md) (if an e2e test has been added)
- [x] I have run `make fmt` and formatted my code

## Further comments

* While loading the config, if the file doesn't exist, check the old path` ~/.config `and load the file copying to the new path `~/.config/mongocli/config.toml`. 
* ~Note: Still trying to figure out linux test failure - but sending for early review~
* ~Note: some evergreen files need to be updated to address this, but I want to run an e2e test without the change to see this solution working.~

### CODE REVIEW

File: internal/config/profile.go
Comment (gssbzn) [inline internal/config/profile.go:631]: [nit] let's not do this here, let's follow a similar approach to what was done for atlas cli, atlas cli will read if a mongocli config exists and copy it to the atlas cli config path, this check is done at run time when calling a command,
Also notice that logic may need to be updated as well to account for two possible paths

Comment (blva) [reply internal/config/profile.go:631]: Do you mean [this](https://github.com/mongodb/mongocli/blob/098509e33af5bba51b4429bcaae7961231c5c15e/cmd/atlas/atlas.go#L59)? 

I initially thought of following the same approach but overall I thought the logic would look a bit complex if I followed that approach.

Since we load the profile at the beginning of the command, we get an error if we update `MongoCLIConfigHome()` ->` Error: credentials not set`. 

Maybe you thought about something else but what I thought was that we would need to create another profile `oldProfile` and a `newProfile` and try loading `newProfile`, then try `oldProfile` if it doesn't exist until one of them exists. If oldProfile exists we return it and then the atlas command will remove the file, making oldProfile invalid and newProfile loadable. So my understanding was that in the end we would need changes in both files and more clutter to make it work. Again this was what I thought as a solution, please let me know if there is a better way I did not see.

File: internal/config/profile.go
Comment (andreaangiolillo) [inline internal/config/profile.go:682]: Could you add deprecatation comments saying that mongocli version prior v1.24.0 uses this path and that new versions use `MongoCLIConfigHome`? 

I'd also rename the function as `DeprecatedMongoCLIConfigHome` but this is just my preference

Comment (andreaangiolillo) [reply internal/config/profile.go:682]: > add a paragraph to its doc comment that begins with “Deprecated:” followed by some information about the deprecation.

https://go.dev/blog/godoc

File: internal/config/profile.go
Comment (andreaangiolillo) [reply internal/config/profile.go:631]: +1

Comment (gssbzn) [reply internal/config/profile.go:631]: >Do you mean [this](https://github.com/mongodb/mongocli/blob/098509e33af5bba51b4429bcaae7961231c5c15e/cmd/atlas/atlas.go#L59)?

yes

> Since we load the profile at the beginning of the command, we get an error if we update MongoCLIConfigHome() -> Error: credentials not set.

The logic above runs before initialziing the profile, it's actually nicely seen here
```
	cobra.OnInitialize(createConfigFromMongoCLIConfig, initConfig)
```
first copy then load config

> I thought was that we would need to create another profile oldProfile and a newProfile and try loading newProfile, then try oldProfile if it doesn't exist until one of them exists

Too many responsibilities for the profile to have to remember there's actually two files, Andrea and I already went down discussing the approach for the linked piece of code, you can revisit [here](https://github.com/mongodb/mongocli/pull/977#discussion_r804860498)

And just in case the linked discussion/approach keep the existing mongocli config, we don't need that part we can move it without keeping a back up. This is also why profile should not have a fallback logic, the entry point should always make sure the file is in the right place 

@andreaangiolillo you may want to expand your `+1` as it's not clear what idea are you supporting here

Comment (andreaangiolillo) [reply internal/config/profile.go:631]: I added my comment when Bianca's reply wasn’t there 🤷‍♂️ .

I agree that we should use the same logic that we are using with atlascli

Comment (blva) [reply internal/config/profile.go:631]: cool, thanks for the explanation! let me try to draft this

File: cmd/atlas/atlas.go
Comment (blva) [inline cmd/atlas/atlas.go:102]: pls check if this makes sense

File: build/ci/set-up-cloud-manager.sh
Comment (gssbzn) [inline build/ci/set-up-cloud-manager.sh:21]: [q] what about the folder?

File: build/ci/evergreen.yml
Comment (gssbzn) [inline build/ci/evergreen.yml:347]: [q] what about the folder?

File: build/ci/set-up-ops-manager.sh
Comment (gssbzn) [inline build/ci/set-up-ops-manager.sh:78]: [q] what about the folder?

File: internal/config/profile.go
Comment (andreaangiolillo) [inline internal/config/profile.go:537]: FYI: we were storing the atlasCLI config in the path `.config/atlascli/atlascli.toml` instead of `.config/atlascli/config.toml`. This should fix the issue 💯

File: build/ci/set-up-ops-manager.sh
Comment (blva) [reply build/ci/set-up-ops-manager.sh:78]: It looks like when `XDG_CONFIG_HOME` is set for atlascli [we don't try to add the folder to the path ](https://github.com/mongodb/mongocli/blob/f319c546a8f66a1f6d2970da6e6127d758e180f0/internal/config/profile.go#L661). So I assumed that for mcli would be the same. Is that something we want to do for both?

File: cmd/atlas/atlas.go
Comment (andreaangiolillo) [inline cmd/atlas/atlas.go:107]: [nit] you can remove these vars as they are already defined

File: cmd/atlas/atlas.go
Comment (andreaangiolillo) [inline cmd/atlas/atlas.go:112]:
```go
	if configDir, err := config.MongoCLIConfigHome(); err == nil{
		configPath = fmt.Sprintf("%s/config.toml", configDir)
	}
```

File: cmd/atlas/atlas.go
Comment (andreaangiolillo) [inline cmd/atlas/atlas.go:117]:
```suggestion
	if oldConfigDir, err := config.OldMongoCLIConfigHome(); err == nil {
		oldConfigPath = fmt.Sprintf("%s/mongocli.toml", oldConfigDir)
	}
```

File: cmd/mongocli/mongocli.go
Comment (gssbzn) [inline cmd/mongocli/mongocli.go:54]: you're leaking these file since you'll never close it

File: cmd/mongocli/mongocli.go
Comment (gssbzn) [inline cmd/mongocli/mongocli.go:49]: I'm just realising this, but, does this work for windows or do we need to use `filepath.Join`?

File: cmd/mongocli/mongocli.go
Comment (andreaangiolillo) [inline cmd/mongocli/mongocli.go:71]: you are already trying to open the old config in line 62 `in, err := os.Open(oldMongoCLIConfigPath)`, if you reach this point the dir should be already there, right?

File: cmd/mongocli/mongocli.go
Comment (andreaangiolillo) [inline cmd/mongocli/mongocli.go:76]: I am not sure why you are checking for the `oldMongoCLIConfigHome` and then create the `mongoCLIConfigHome` if the old dir is not there

File: build/ci/set-up-ops-manager.sh
Comment (gssbzn) [reply build/ci/set-up-ops-manager.sh:78]: oh I think this uncovers a bug on the atlas implementation when `XDG_CONFIG_HOME` is set the config path for both atlas and now mongocli should be `XDG_CONFIG_HOME/{mongocli|atlascli}/config.toml`

File: cmd/atlas/atlas.go
Comment (gssbzn) [inline cmd/atlas/atlas.go:105]: is this method necessary? you're probably only gonna use one, can we just call them separately instead of a method to get both at the same time?

File: cmd/atlas/atlas.go
Comment (gssbzn) [inline cmd/atlas/atlas.go:139]: `return os.Open(oldMongoCLIConfigPath)`

File: cmd/mongocli/mongocli.go
Comment (blva) [reply cmd/mongocli/mongocli.go:76]: my bad, this should be `mongoCLIConfigHome`, thank you!

File: build/ci/evergreen.yml
Comment (blva) [reply build/ci/evergreen.yml:347]: fixing

File: cmd/atlas/atlas.go
Comment (blva) [reply cmd/atlas/atlas.go:105]: sure!

File: cmd/atlas/atlas.go
Comment (blva) [reply cmd/atlas/atlas.go:107]: I'll still need it since I'll move this to another method now. Please, LMK if you have another suggestion in the next rev

File: cmd/mongocli/mongocli.go
Comment (blva) [reply cmd/mongocli/mongocli.go:49]: I would assume it works since `MongoCLIConfigHome` has been using this already? How can I confirm if this works?

File: cmd/mongocli/mongocli.go
Comment (blva) [reply cmd/mongocli/mongocli.go:71]: yes, this was the wrong one as you pointed in your other comment, thx :)

File: cmd/mongocli/mongocli.go
Comment (andreaangiolillo) [reply cmd/mongocli/mongocli.go:49]: 1. spawn a windows host on evergreen
2. run the task package mongocli msi on evergreen
3. access your windows host with Microsoft Remote Desktop
4. install the msi created at step 2, on your windows machine

Comment (blva) [reply cmd/mongocli/mongocli.go:49]: Thanks! On it

File: internal/config/profile.go
Comment (fmenezes) [inline internal/config/profile.go:662]: I don't know if that was agreed already on a spec but a more OS compatible way for the .config folder is [os.UserConfigDir](https://pkg.go.dev/os#UserConfigDir). Currently our code always defaults to `os.UserHomeDir() / .config` but the above mentioned code would try `$HOME/Library/Application Support` for macos and `%AppData%` for Windows which is probably what our customers would have expected.

Not blocking just mentioning here since we are changing folders for the CLI this might by the right moment to rethink this folder.

Comment (andreaangiolillo) [reply internal/config/profile.go:662]: I didn’t know about this. Thanks for sharing it! I am in favor of this change as we are already changing the location of the config. Tagging @gssbzn to see if we all agree on this change.

Comment (blva) [reply internal/config/profile.go:662]: sounds great, to see if I understood -  we would replace the whole logic within `MongoCLIConfigHome() `and use `os.UserConfigDir()`, furthermore we would no longer use `.config`. Just simply create a mongocli folder under `UserConfigDir`. Adding to this PR sounds reasonable.

File: cmd/mongocli/mongocli.go
Comment (blva) [reply cmd/mongocli/mongocli.go:49]: Ok, tested and we're creating the right folder structure if that's the concern here:

```
PS C:\Users\Administrator\AppData\Roaming> ls


    Directory: C:\Users\Administrator\AppData\Roaming


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----        1/31/2022   5:40 PM                ActiveState
d-----         3/7/2022   4:06 PM                Adobe
d---s-         3/7/2022   4:06 PM                Microsoft
d-----         3/7/2022   4:21 PM                mongocli
d-----        1/31/2022   5:57 PM                npm
d-----        1/31/2022   5:31 PM                NuGet
d-----        1/31/2022   5:51 PM                Visual Studio Setup

```
```
PS C:\Users\Administrator\AppData\Roaming> cd mongocli
PS C:\Users\Administrator\AppData\Roaming\mongocli> ls


    Directory: C:\Users\Administrator\AppData\Roaming\mongocli


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----         3/7/2022   4:21 PM             45 config.toml

```
```
PS C:\Users\Administrator\AppData\Roaming\mongocli> cat config.toml

[default]
  output = ""
  service = "cloud"
```

File: internal/config/profile.go
Comment (blva) [reply internal/config/profile.go]: I added the changes for this, LMK if we can proceed with this approach.

File: internal/config/profile.go
Comment (fmenezes) [inline internal/config/profile.go:676]: Looks like in `MongoCLIConfigHome` there is no fallback but in `AtlasCLIConfigHome` we do have a fallback to `UserHomeDir`, any reason for the difference?

I'd be inclined to keep things in sync.

Comment (fmenezes) [reply internal/config/profile.go:676]: Without a fallback

File: internal/config/profile.go
Comment (fmenezes) [inline internal/config/profile.go:648]: [q] Should we really return an error here?

File: internal/config/profile.go
Comment (blva) [reply internal/config/profile.go:676]: no reason other than me me forgetting to remove the fallback, synced w/ you offline that is OK to remove for AtlasCLIConfigHome

File: internal/config/profile.go
Comment (blva) [reply internal/config/profile.go:648]: not anymore since we're now changing this as well. thanks! updating

File: cmd/atlas/atlas.go
Comment (fmenezes) [inline cmd/atlas/atlas.go:125]: nitpick: my personal style is never to open files within functions like this rather to return the path and let the consumer open/close files, I personally believe writing code to open/close should always be together (that is one of the strengths of defer keyword).

If you are relying on open errors to fallback consider:
```go
if _, err := os.Stat("/path/to/whatever"); errors.Is(err, os.ErrNotExist) {
  // path/to/whatever does not exist
}
```

**Review by fmenezes — APPROVED** [review]

File: cmd/atlas/atlas.go
Comment (gssbzn) [inline cmd/atlas/atlas.go:130]:
```suggestion
	return os.Stat(configPath)
```

File: cmd/atlas/atlas.go
Comment (gssbzn) [inline cmd/atlas/atlas.go:123]:
```suggestion
	if configDir, err := config.OldMongoCLIConfigHome(); err == nil {
```

File: cmd/mongocli/mongocli.go
Comment (gssbzn) [reply cmd/mongocli/mongocli.go:49]: that looks like powershell, I may give this a try in `cmd` but let's not block the PR on this

File: cmd/mongocli/mongocli.go
Comment (gssbzn) [inline cmd/mongocli/mongocli.go:87]:
```suggestion
	if _, err = io.Copy(out, in); err != nil {
```

File: cmd/atlas/atlas.go
Comment (blva) [reply cmd/atlas/atlas.go:130]: this returns FileInfo object and we're returning file path here? 
```
type FileInfo interface {
	Name() string       // base name of the file
	Size() int64        // length in bytes for regular files; system-dependent for others
	Mode() FileMode     // file mode bits
	ModTime() time.Time // modification time
	IsDir() bool        // abbreviation for Mode().IsDir()
	Sys() interface{}   // underlying data source (can return nil)
}
```

**Review by gssbzn — DISMISSED** [review]
LGTM thanks for all the work here, none of my comments are blocking and are mostly style

Let's not merge this until 1.23.1 is out as this is a big change for a bug fix version

File: cmd/atlas/atlas.go
Comment (gssbzn) [reply cmd/atlas/atlas.go:130]: true, my bad

**Review by andreaangiolillo — APPROVED** [review]
LGTM. Thanks for addressing all the comments! I have selected additional evergreen tasks to double-check that everything works as expected. Okay to merge this once we release 1.23.1 and all tasks are successful.

**Comment by andreaangiolillo** [comment]
CM tests are failing because of one of the changes in this PR https://evergreen.mongodb.com/task_log_raw/mongocli_master_e2e_cloud_manager_remote_cloud_manager_deploy_replica_set_e2e_patch_2911a9e4aba35f76cfa2ab03053c9510c83f47d1_62264cc20305b9056d333730_22_03_07_18_19_47/0?type=T#L47

**Review by andreaangiolillo — APPROVED** [review]
LGTM if tests are successful. Thanks!

**Pull request merged**
